src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs

```diff
@@ -298,19 +343,33 @@ public void RunClassInitialize(TestContext testContext)
         /// <returns>
         /// Any exception that can be thrown as part of a class cleanup as warning messages.
         /// </returns>
-        [SuppressMessageAttribute("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Requirement is to handle all kinds of user exceptions and message appropriately.")]
+        [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Requirement is to handle all kinds of user exceptions and message appropriately.")]
         public string RunClassCleanup()
         {
-            if (this.ClassCleanupMethod == null)
+            if (this.ClassCleanupMethod is null)
             {
-                return null;
+                if (this.BaseClassCleanupMethodsQueue.Count == 0)
+                {
+                    return null;
+                }
             }
 
-            if (this.IsClassInitializeExecuted || this.ClassInitializeMethod == null)
+            if (this.IsClassInitializeExecuted || this.ClassInitializeMethod is null || this.BaseClassInitializeMethodsDict.ContainsValue(true))
             {
+                var classCleanupMethod = this.classCleanupMethod;
                 try
                 {
-                    this.ClassCleanupMethod.InvokeAsSynchronousTask(null);
+                    classCleanupMethod?.InvokeAsSynchronousTask(null);
+                    var baseClassCleanupQueue = new Queue<MethodInfo>(this.BaseClassCleanupMethodsQueue);
+                    while (baseClassCleanupQueue.Count > 0)
+                    {
+                        classCleanupMethod = baseClassCleanupQueue.Dequeue();
+                        if (this.BaseClassInitializeMethodsDict.Any(method => method.Key.DeclaringType
+                                == classCleanupMethod.DeclaringType && method.Value))
```

Comment:
This would probably not work in the following case. We wouldn't call the base class cleanup cause we haven't run its initialize.
```
class A
{
 [ClassCleanup]
public static void CC() {}
}

public class B : A
{
[ClassInitialize]
public static void DCI(TestContext tc) {}
[ClassCleanup]
public static void DCC() {}
}
```

Reply:
I like that we aren't running cleanup for initializes that haven't happened but this would probably only work correctly if we keep track of the base types hierarchy.

Reply:
It might be ok to just run Cleanups blind like TestCleanups to start with. Not my favorite choice but should be ok to start with.

Reply:
well, not mine either honestly. I think that in order to be symmetrical with initialize the clean up should only be executed when its initialize has happened. 

I think that the if check:
`this.ExecutedBaseClassInitializeMethods.Any(method => method.DeclaringType == classCleanupMethod.DeclaringType)` will only let cleanup to execute if both initialize and cleanup method share same declaring type.

I'm open to make the change either way. Let me know what you think.

Reply:
I have another idea - how about storing init and cleanup as tuples? That way cases like the ones above can be accounted for as well. You'd only execute cleanups whose corresponding initialize tuple has been executed without any exceptions.

Reply:
it sounds good. one question, how different would be using a dictionary instead of a tuple? will we gain any advantage for using it?

Reply:
So the tuple would help you associate Init,Cleanup pairs so cases like the above where you have a cleanup but no init would still work(you;d just push in a null init with the cleanup). You'd probably still need some form a linear data structure(a stack should work) to determine what cleanups to run.

Reply:
hey @AbhitejJohn 
latest changes i pushed cover what we discussed through the review comments.
Please, take a look at the code and let me know if further changes are required, otherwise, I think we should be good.